### ISSUE TITLE
update the onnxruntime-web version to support the multithreads

### ISSUE THREAD

**Issue by helloburke** [huggingface/transformers.js#882]
### Feature request

// Due to a bug in onnxruntime-web, we must disable multithreading for now.
// See [[Web] chrome V3 extension `TypeError: URL.createObjectURL is not a function` · Issue #14445 · micros](https://github.com/microsoft/onnxruntime/issues/14445) for more information.
env.backends.onnx.wasm.numThreads = 1;

the  issue of https://github.com/microsoft/onnxruntime/issues/14445 was closed. 
Is it possible to enable multi-threading for onnxruntime-web with WASM?

### Motivation

update the onnxruntime-web version to support the multithreads

### Your contribution

no

**Comment by kungfooman** [comment]
PR for the issue you mentioned:

https://github.com/microsoft/onnxruntime/pull/20165#issue-2217881630

Looks like it's at: [onnxruntime-web@1.19.0-esmtest.20240513-a16cd2bd21](https://www.npmjs.com/package/onnxruntime-web/v/1.19.0-esmtest.20240513-a16cd2bd21)

Even [V3 branch](https://github.com/xenova/transformers.js/pull/545) is still at `1.18.0` and it also contains still references to `1.17.1`:

```js
env.backends.onnx.wasm.wasmPaths = 'https://cdn.jsdelivr.net/npm/onnxruntime-web@1.17.1/dist/';
env.backends.onnx.wasm.numThreads = 1;
```

Based on semver I think you should be able to just install the `1.19` package, but the PR sounded like a little bigger refactor, so I wouldn't be surprised if we need more changes for this. You wanna test this and report?

**Comment by fs-eire** [comment]
latest v3 branch is using `"onnxruntime-web": "1.19.0-dev.20240804-ee2fe87e2d",` and I believe WebAssembly multi-thread should be already supported in this version (unless you use in CSP restricted environment, eg. ServiceWorker). Please have a try and let me know if it's still not working.

**Comment by kungfooman** [comment]
Yes, your PR was merged a day after I mentioned it: https://github.com/xenova/transformers.js/commit/437cb34e50a27e40237c3a8eee3527b2db459d58

And the `wasmPaths` referring to `1.17.1` are still not updated, which may cause issues with those two examples regarding `numThreads`.

I wanted to try it, but the amount of magic spells you have to cast (aka Chrome “security” requirements) to make it work is hilarious.

I enabled this option:

![image](https://github.com/user-attachments/assets/bb10726e-44bf-4cea-8cbd-47f614d9e4e8)

And I don't get a `SharedArrayBuffer` anyway:

![image](https://github.com/user-attachments/assets/b2c6fd5e-0dda-4b07-a137-e5ff827faec1)

And ONNX only seems to test `self.crossOriginIsolated` which is `false` without wasting more time on random server settings.

**Comment by flatsiedatsie** [comment]
Cool, nice experiment. And it's exactly the model I was hoping to speed up, as the rest of my voice-that-with-an-AI pipeline, namely Whisper and LLM's, enjoys some level of GPU speedup.

Would this work without casting magic spells on MacOS or Windows? Or is this still behind a flag on all browsers?

According to [Can I Use](https://caniuse.com/sharedarraybuffer) sharedArrayBuffers are widely implemented?

**Comment by kungfooman** [comment]
> Would this work without casting magic spells on MacOS or Windows? Or is this still behind a flag on all browsers?

If you send these two headers via PHP or some other server setuo to enable cross-origin isolation it might work on all OS'es (I just couldn't be bothered to spawn a special server just for a quick test). But yea, why is it even still "experimental" then? Only a test will tell, that you might wanna do?

**Comment by flatsiedatsie** [comment]
Yes, I'm already on it. I'm sure I have cross-origin isolation through htaccess.

Fun fact: you can actually create cross-origin isolationn via a service worker, no serverside adjustments needed: https://github.com/gzuidhof/coi-serviceworker

**Comment by flatsiedatsie** [comment]
OK, I've run a test.

It seems that by default Transformers.js already uses multi-threading. What's more, it seems to already have picked the optimum number (which I often read is the number of threads divided by two). This bears out in the tests:

1 -> 9 seconds
2 -> 5.7 seconds
4 -> 3.5 seconds (automatic setting)
8 -> 4.5 seconds (my hardwareConcurrency)

Details:
This is on a Macbook Pro M1 with 16Gb of ram, using the Brave browser.

1 thread
<img width="624" alt="Screenshot 2024-08-15 at 13 56 07" src="https://github.com/user-attachments/assets/83da1069-5708-442c-a8ee-a734a5dfc9f8">


2 threads
<img width="622" alt="Screenshot 2024-08-15 at 14 02 15" src="https://github.com/user-attachments/assets/b38369bc-4028-4482-bdb4-1da77f3e3c14">


4 threads
<img width="623" alt="Screenshot 2024-08-15 at 13 54 25" src="https://github.com/user-attachments/assets/6b432c63-04ab-466a-83a3-5c4da51e01e8">


8 threads
<img width="611" alt="Screenshot 2024-08-15 at 13 52 59" src="https://github.com/user-attachments/assets/846a6c56-c802-4224-84eb-4e6b322c57da">


My settings:
```
static async getInstance(progress_callback = null) {
        if (this.tokenizer_instance === null) {
            this.tokenizer = AutoTokenizer.from_pretrained(this.model_id, { progress_callback });
        }

        if (this.model_instance === null) {
            this.model_instance = SpeechT5ForTextToSpeech.from_pretrained(this.model_id, {
                //quantized: false,
				dtype: 'fp32',
				quantized:true,
                progress_callback,
				device:self.device,
            });
        }
        if (this.vocoder_instance === null) {
            this.vocoder_instance = SpeechT5HifiGan.from_pretrained(this.vocoder_id, {
                //quantized: false,
				dtype: 'fp32',
				quantized:true,
                progress_callback,
				device:self.device,
            });
        }

        return new Promise(async (resolve, reject) => {
            const result = await Promise.all([
                this.tokenizer,
                this.model_instance,
                this.vocoder_instance,
            ]);
            self.postMessage({
                status: 'ready',
            });
            resolve(result);
        });
    }

```

**Comment by xenova** [comment]
Closing as this was implemented a while ago :) you can also try it in our preview release of Transformers.js v4 (https://github.com/huggingface/transformers.js/pull/1382)
